### ISSUE TITLE
Event tracking Recorder

### ISSUE THREAD

**Issue by jetuk** [pywr/pywr#170]
We need a generic `Recorder` for tracking event statistics. This should include some sort of `Parameter` based threshold for which statistics such as number and duration of breaches are recorded.

**Comment by snorfalorpagus** [comment]
The threshold would be useful not just for a recorder, but also some parameters. e.g. is the average flow for the last 10 days above X? see #155.

**Comment by jetuk** [comment]
I was thinking the threshold could be any `Parameter` object.

**Comment by snorfalorpagus** [comment]
What does the output of this recorder look like? How about a dataframe with a row for each "event" with a column for start and end timestep. Should we recorder the timestep or the timestamp?

**Comment by jetuk** [comment]
Both? It should also record something like the min, max and mean values during the event.

**Comment by snorfalorpagus** [comment]
> I was thinking the threshold could be any Parameter object. However this might not work with the way Recorder saves data.

I can think of three things we'd want to track:

- The flow at a node
- The storage at a node
- The value of a parameter

We already have recorders that capture the flow and storage. Capturing a parameter shouldn't be too difficult (loop through the scenarios in `save` and call `value`).

The clever bit comes in how the events are logged. Rememeber that events might have different durations in different scenarios. Is the output a list of data frames?

If you're interested in min/mean/max during an event, presumably the thing you're measuring might be distinct from the event trigger itself? e.g. an event might be demand saving triggering for which you want to record the duration, but you might be interested in... something else?

**Comment by snorfalorpagus** [comment]
We discussed this today. We decided that as a minimum the recorder should save the start and end timestep of the event. Each scenario's events are separate. We also talked about tracking user-defined parameters during the event, e.g. during the event the minimum storage of reservoir 1 was X, reservoir 2 was Y. This would fit nicely in a table:

```
EID, SID, START, END, RES1, RES2
0, 0, 1920-01-01, 1920-01-05, 23424, 3453
```

We then have the option to save this information directly to disk (pytables) or just keep it in memory (a dataframe).
